waku/v2/protocol/waku_noise/noise.nim

```diff
@@ -98,10 +261,333 @@ proc genKeyPair*(rng: var BrHmacDrbgContext): KeyPair =
+# The CipherState is updated increasing the nonce (used as a counter in Noise) by one
+proc decryptWithAd*(state: var CipherState, ad, ciphertext: openArray[byte]): seq[byte]
+  {.raises: [Defect, NoiseDecryptTagError, NoiseNonceMaxError].} =
+
+  # We read the authorization appendend at the end of a ciphertext
+  let inputAuthorizationTag = ciphertext.toOpenArray(ciphertext.len - ChaChaPolyTag.len, ciphertext.high).intoChaChaPolyTag
+    
+  var
+    authorizationTag: ChaChaPolyTag
+    nonce: ChaChaPolyNonce
+
+  # The nonce is read from the input CipherState
+  # By Noise specification the nonce is 8 bytes long out of the 12 bytes supported by ChaChaPoly
+  nonce[4..<12] = toBytesLE(state.n)
+
+  # Since ChaChaPoly decryption primitive overwrites the input with the output,
+  # we copy the ciphertext (authorization tag excluded) in the output plaintext variable and we pass it to decryption  
+  var plaintext = ciphertext[0..(ciphertext.high - ChaChaPolyTag.len)]
+  
+  ChaChaPoly.decrypt(state.k, nonce, authorizationTag, plaintext, ad)
+
+  # We check if the input authorization tag matches the decryption authorization tag
+  if inputAuthorizationTag != authorizationTag:
+    debug "decryptWithAd failed", plaintext = plaintext, ciphertext = ciphertext, inputAuthorizationTag = inputAuthorizationTag, authorizationTag = authorizationTag
+    raise newException(NoiseDecryptTagError, "decryptWithAd failed tag authentication.")
+  
+  # We increase the Cipher state nonce
+  inc state.n
+  # If the nonce is greater than the maximum allowed nonce, we raise an exception
+  if state.n > NonceMax:
+    raise newException(NoiseNonceMaxError, "Noise max nonce value reached")
+
+  trace "decryptWithAd", inputAuthorizationTag = inputAuthorizationTag, authorizationTag = authorizationTag, nonce = state.n
+    
+  return plaintext
+
+# Sets the nonce of a Cipher State
+proc setNonce*(cs: var CipherState, nonce: uint64) =
+  cs.n = nonce
+
+# Sets the key of a Cipher State
+proc setCipherStateKey*(cs: var CipherState, key: ChaChaPolyKey) =
+  cs.k = key
+
+# Generates a random Symmetric Cipher State for test purposes
+proc randomCipherState*(rng: var BrHmacDrbgContext, nonce: uint64 = 0): CipherState =
+  var randomCipherState: CipherState
+  brHmacDrbgGenerate(rng, randomCipherState.k)
+  setNonce(randomCipherState, nonce)
+  return randomCipherState
+
+
+# Gets the key of a Cipher State
+proc getKey*(cs: CipherState): ChaChaPolyKey =
+  return cs.k
+
+# Gets the nonce of a Cipher State
+proc getNonce*(cs: CipherState): uint64 =
+  return cs.n
+
+#################################
+# Symmetric State primitives
+#################################
+
+# Initializes a Symmetric State
+proc init*(_: type[SymmetricState], hsPattern: HandshakePattern): SymmetricState =
+  var ss: SymmetricState
+  # We compute the hash of the protocol name
+  ss.h = hsPattern.name.hashProtocol
+  # We initialize the chaining key ck
+  ss.ck = result.h.data.intoChaChaPolyKey
```

Comment:
Great catch! This bug was setting the initial chaining key to an empty value, rather than the handshake hash value (thus losing the initial binding with the protocol name). I fixed and added an extra check in unit tests to see if `h` and `ck` agree on their when initializing a Symmetric State.

Reply:
Addressed in https://github.com/status-im/nwaku/pull/933/commits/4ae4364845643190a3af13b3f15fd4afb6bcf906